### PR TITLE
Bug 1920205: use secure local registry for e2e tests

### PR DESCRIPTION
**Pull request by exdx**
**Description of the change:**
Use local docker registry instead of quay.io for pushing images during the course of opm e2e tests

**Motivation for the change:**
The e2e tests are currently skipped on forks because the quay credentials cannot be securely provided to the fork. The e2e image has been made public and a local registry is spun up during the course of the test instead. 

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

**Comment by exdx** [comment]
The e2e tests are now passing -- all tests are being run with podman, docker, and containerd enabled. The goal of the PR was to enable testing against a local registry without make any substantial changes to the tests themselves. 

The only change to code was the potential for a data race that was uncovered in the containerd resolver. 
```
 WARNING: DATA RACE
Write at 0x000003cea580 by goroutine 39:
  github.com/operator-framework/operator-registry/pkg/image/containerdregistry.NewResolver()
      /home/runner/work/operator-registry/operator-registry/pkg/image/containerdregistry/resolver.go:44 +0x51b
  github.com/operator-framework/operator-registry/pkg/image/containerdregistry.NewRegistry()
      /home/runner/work/operator-registry/operator-registry/pkg/image/containerdregistry/options.go:93 +0x69c
  github.com/operator-framework/operator-registry/pkg/lib/bundle.(*BundleExporter).Export()
      /home/runner/work/operator-registry/operator-registry/pkg/lib/bundle/exporter.go:47 +0x7bd
  github.com/operator-framework/operator-registry/pkg/lib/indexer.ImageIndexer.ExportFromIndex.func1()
      /home/runner/work/operator-registry/operator-registry/pkg/lib/indexer/indexer.go:547 +0x266
Previous write at 0x000003cea580 by goroutine 34:
  github.com/operator-framework/operator-registry/pkg/image/containerdregistry.NewResolver()
      /home/runner/work/operator-registry/operator-registry/pkg/image/containerdregistry/resolver.go:44 +0x51b
  github.com/operator-framework/operator-registry/pkg/image/containerdregistry.NewRegistry()
      /home/runner/work/operator-registry/operator-registry/pkg/image/containerdregistry/options.go:93 +0x69c
  github.com/operator-framework/operator-registry/pkg/lib/bundle.(*BundleExporter).Export()
      /home/runner/work/operator-registry/operator-registry/pkg/lib/bundle/exporter.go:47 +0x7bd
  github.com/operator-framework/operator-registry/pkg/lib/indexer.ImageIndexer.ExportFromIndex.func1()
      /home/runner/work/operator-registry/operator-registry/pkg/lib/indexer/indexer.go:547 +0x266
```

Having this PR in the codebase will enable registry e2e tests to run on every commit and help ensure stability for large changes to the registry like in the declarative index proposal.

File: test/e2e/e2e_suite_test.go [hunk 1]
```diff
@@ -20,13 +22,27 @@ func TestE2E(t *testing.T) {
...
+		// no need to login
 		return
 	}
 
-	dockerlogin := exec.Command("docker", "login", "-u", dockerUsername, "-p", dockerPassword, "quay.io")
-	err := dockerlogin.Run()
-	Expect(err).NotTo(HaveOccurred(), "Error logging into quay.io")
+	if dockerHost == "" {
+		// Default to Quay.io
+		dockerHost = "quay.io"
```
Comment (dinhxuanvu) [inline test/e2e/e2e_suite_test.go:39]: nit: can we const this to `DEFAULT_REGISTRY` or something? It is easier to know the info on the top instead of burying somewhere in the middle. Not a big deal though.

**Review by dinhxuanvu — COMMENTED** [review]
Looking good. Just a nit.

File: test/e2e/opm_test.go [hunk 2]
```diff
@@ -428,10 +409,20 @@ var _ = Describe("opm", func() {
 	}
 
 	Context("using docker", func() {
-		IncludeSharedSpecs("docker")
+		if err := exec.Command("docker").Run(); err == nil {
+			IncludeSharedSpecs("docker")
+		} else {
+			GinkgoT().Logf("container tool docker not found - skipping docker-based opm e2e tests: %s", err)
+			return
+		}
```
Comment (timflannagan1) [inline test/e2e/opm_test.go:417]: Just a quick drive-by, but you could deflate this to a single conditional by reversing this logic and exiting early.

File: test/e2e/opm_test.go [hunk 3]
```diff
@@ -359,31 +352,19 @@ var _ = Describe("opm", func() {
...
-			workingDir, err := os.Getwd()
-			Expect(err).NotTo(HaveOccurred())
-			err = os.Remove(workingDir + "/" + bundle.DockerFile)
+			err := buildIndexWith(containerTool, "", indexImage, []string{}, registry.ReplacesMode, true)
 			Expect(err).NotTo(HaveOccurred())
 		})
 
-		It("can overwrite existing bundles in an index", func() {
-
+		PIt("can overwrite existing bundles in an index", func() {
```
Comment (dinhxuanvu) [inline test/e2e/opm_test.go:366]: Is this a leftover or intentional?

**Review by dinhxuanvu — COMMENTED** [review]
Just a quick question.

**Review by dinhxuanvu — APPROVED** [review]
/lgtm
